daemon/graphdriver/overlay2/overlay.go

```diff
@@ -230,6 +232,15 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 
 	logger.Debugf("backingFs=%s,  projectQuotaSupported=%v", backingFs, projectQuotaSupported)
 
+	// figure out whether "index=off" option is recognized by the kernel
+	_, err = os.Stat("/sys/module/overlay/parameters/index")
+	if !os.IsNotExist(err) {
+		// in case Stat() returned any other error than ENOENT,
+		// there is no way to figure out whether index= option
+		// is supported or not, so let's assume it is.
```

Comment:
What's the result if it's not supported and we set this option? Will it fail?

Reply:
Yes, it will most probably result in kernel returning `EINVAL` from `mount()` syscall. This will only happen if:
 * kernel is > 4.13rc1 (and the relevant feature is not backported);
and
 * for some reason /sys/module/overlay/parameters/index is present but `stat()` returns an error.

The former is quite possible, say on older distros.

The latter is highly unlikely to happen, from the top of my head I can only think of a severe kernel memory shortage, or using any security mechanisms (say selinux) to disallow dockerd access to e.g. /sys/module/ directory.

Reply:
Right, so I was thinking if we should take the safe approach, and only add `index=off` if we're successfully able to detect that it's supported (i.e. `os.Stat()` was successfull), so:

- `os.Stat()` was successful; add `index=off`
- `os.Stat()` produces an `IsNotExist`: don't add `index=off`, and continue silently (as this is an expected situation)
- `os.Stat()` produces any other error: log a warning (`"unable to detect if index=off is supported: <original error message>"`), but continue as usual (but _don't_ add `index=off`)

Reply:
Logging a warning at least makes it visible that we were unable to detect, and may assist in troubleshooting why we failed to detect.

Reply:
Makes sense; please see updated commit